### PR TITLE
[VIS] Role visualisation

### PR DESCRIPTION
**Pull request by olly-larkin**
# Summary

Added role visualisation. Updates for new runs (currently no change seen as voting does not occur).

image

## Additional Information

Currently using old version of recharts as newest has bug which causes compilation errors.

## Test Plan

If disabling using new runs and modify initial `output.json`, changes can be seen after refreshing the page.

image

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

**Comment by lhl2617** [comment]
image
nice

File: website/src/components/Visualisations/Roles/Roles.module.css [hunk 1]
```diff
@@ -5,29 +5,25 @@
 
 .root {
     margin: 0;
+    width: 50%;
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Roles.module.css:8]: this might not look good on mobile

Comment (lhl2617) [reply website/src/components/Visualisations/Roles/Roles.module.css:8]: but that's really a stretch goal, so nvm. you can however try `min-width: 50%` and see how it looks like

File: website/src/components/Visualisations/Roles/Roles.tsx [hunk 2]
```diff
@@ -1,14 +1,174 @@
...
+    Bar,
+    XAxis,
+    YAxis,
+    Tooltip,
+    Legend,
+    ResponsiveContainer,
+} from "recharts";
+import { TurnsInRoles, ProcessedRoleData } from "./Util/RoleTypes";
+import { getProcessedCIRoleData, processRoleData } from "./Util/ProcessedRoleData";
+import { loadLocalOutput } from "../../NewRun/utils";
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Roles.tsx]: would need to refactor after #230 goes in

File: website/src/components/Visualisations/Roles/Roles.tsx [hunk 3]
```diff
@@ -1,14 +1,174 @@
...
+            turnsInRoles.judge +
+            turnsInRoles.speaker +
+            turnsInRoles.none;
+
+        return (
+            <div className={styles.customTooltip}>
+                <p className={styles.label}>{label}</p>
+                <p className={styles.content}>
+                    Turns as President: {turnsInRoles.president} (
+                    {((turnsInRoles.president * 100) / turns).toFixed(1)}%)
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Roles.tsx]: what is this 100 magik numba?

File: website/src/components/Visualisations/Roles/Roles.tsx [hunk 4]
```diff
@@ -1,14 +1,174 @@
...
+                </p>
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Roles.tsx]: fine but you can use some sik tricks to map from `[president, judge, speaker, none].map(....// sik trick here)`

File: website/src/components/Visualisations/Roles/Roles.tsx [hunk 5]
```diff
@@ -1,14 +1,174 @@
...
+};
 
 const Roles = () => {
-  return (
-    <div className={styles.root}>
-      <img src={logo} className={styles.appLogo} alt="logo" />
-      <p className={styles.text}>Role Visualisation</p>
-    </div>
-  );
+    const presidentColor = "#00bbf9";
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Roles.tsx]: mmm  lots of hardcoded roles so might be cumbersome if we add roles, but we won't--so this is fine.

File: website/src/components/Visualisations/Roles/Roles.tsx [hunk 6]
```diff
@@ -1,14 +1,174 @@
...
+            setData(processRoleData(localOutput.output));
+        }
+    };
+
+    return (
+        <div className={styles.root}>
+            <ResponsiveContainer height={460} width="100%">
+                <BarChart data={data} layout="vertical" margin={{ bottom: 30 }}>
+                    <YAxis type="category" dataKey="name" />
+                    <XAxis type="number" />
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Roles.tsx]: label turns?

File: website/src/components/Visualisations/Roles/Util/RoleTypes.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    constructor(
+        president?: number,
+        judge?: number,
+        speaker?: number,
+        none?: number
+    ) {
+        this.president = president || 0;
+        this.judge = judge || 0;
+        this.speaker = speaker || 0;
+        this.none = none || 0;
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Util/RoleTypes.ts:18]: is 0 Team1's value?

Comment (lhl2617) [reply website/src/components/Visualisations/Roles/Util/RoleTypes.ts:18]: ah - ok, these are num turns in roles.

File: website/src/components/Visualisations/Roles/Util/RoleTypes.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    name: string;
+    roles: TurnsInRoles[];
+
+    constructor(name: string, roles: TurnsInRoles[] = [new TurnsInRoles()]) {
+        this.name = name;
+        this.roles = roles;
+    }
+
+    increment(role: Role) {
+        if (this.roles[this.roles.length - 1].get(role) === 0) {
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Util/RoleTypes.ts]: ~vibe~ bounds check

File: website/src/components/Visualisations/Roles/Util/RoleTypes.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            const newTurnsInRoles = new TurnsInRoles();
+            newTurnsInRoles.increment(role);
+            this.roles.push(newTurnsInRoles);
+        } else {
+            this.roles[this.roles.length - 1].increment(role);
```
Comment (lhl2617) [inline website/src/components/Visualisations/Roles/Util/RoleTypes.ts]: ~vibe~ bounds check

**Review by lhl2617 — CHANGES_REQUESTED** [review]
Looks really good to me, good job. 

Let's merge in #230 and #231 first

File: website/src/components/Visualisations/Roles/Roles.tsx (hunk 3 above)
Comment (QFSW) [reply website/src/components/Visualisations/Roles/Roles.tsx]: Looks like decimal to percentage conversion in which case the magic number is fine

File: website/src/components/Visualisations/Roles/Util/RoleTypes.ts (hunk 7 above)
Comment (lhl2617) [reply website/src/components/Visualisations/Roles/Util/RoleTypes.ts:18]: use `??` btw -- i think 0 || 1 -> 1 (not an issue here but just good pratice)

File: website/src/components/Visualisations/Roles/Roles.module.css (hunk 1 above)
Comment (olly-larkin) [reply website/src/components/Visualisations/Roles/Roles.module.css:8]: Going to leave this for now as I can't see how this graph would work on mobile without rotating (at which point just rotate the device)

**Review by lhl2617 — APPROVED** [review]
this boi clean
image

**Pull request merged**